### ISSUE TITLE
cmd/go: improve error message for unparsable, too-new, go.mod

### ISSUE THREAD

**Issue by prattmic** [golang/go#70979]
Split from #70949.

The tl;dr of that issue is:

1. Set GOTOOLCHAIN=local to prevent automatic toolchain upgrade
2. Use Go 1.23 toolchain
3. go.mod says 1.24rc1 and contains a `tool` block

The 1.23 toolchain does not upgrade because of GOTOOLCHAIN=local, so it attempts to parse the go.mod itself. It does not understand `tool` blocks, so it reports a parse error:

```
go: errors parsing go.mod:
go.mod:206: unknown block type: tool
```

We could have a better error message in this case. Perhaps one of:

1. Add a hint to the above error that GOTOOLCHAIN=local prevented an upgrade
2. Or (my preference), stop immediately after determining the module Go version without fully parsing the go.mod, since we won't build such a module anyway.

**Comment by prattmic** [comment]
Upon further thought, if there is no `tool` block, `GOTOOLCHAIN=local go build` says "go: go.mod requires go >= 1.24rc1 (running go 1.24; GOTOOLCHAIN=local)". IMO the error should be identical with the addition of a `tool` block.
